Worker should exit with ctx.exit to get the right exitcode for non-zero cases

palfrey opened the pull request:
## Description

Fixes #7358

palfrey commented:
> good tests!

It's a good test, but doesn't appear to be getting picked up by the test coverage, and despite some work I can't see a way to make a unit test version. I'm not seeing codecov in the CI for the integration tests. I can see two ways to fix this: add the integration tests to code coverage, or merge with a slightly lowered test coverage level. Thoughts?

auvipy commented:
> > good tests!
> 
> It's a good test, but doesn't appear to be getting picked up by the test coverage, and despite some work I can't see a way to make a unit test version. I'm not seeing codecov in the CI for the integration tests. I can see two ways to fix this: add the integration tests to code coverage, or merge with a slightly lowered test coverage level. Thoughts?

very good points. we do not have extensive integration tests right now. code coverage is small ATM. for unit tests the primary author who refatcor this part of code, left unit test for future which is a over due. as the integration tests are passing we can go with slight lower test coverage.

auvipy commented:
additonally you can try basic test for worker here https://github.com/celery/celery/tree/master/t/unit/bin using https://pypi.org/project/pytest-click/1.1.0/

palfrey commented:
> additonally you can try basic test for worker here https://github.com/celery/celery/tree/master/t/unit/bin using https://pypi.org/project/pytest-click/1.1.0/

Thanks, added one using that. Was seeing one integration test failure, but that's with `t/integration/test_canvas.py::test_chord::test_mutable_errback_called_by_chord_from_group_fail_multiple[errback_old_style]` which I'm pretty sure has nothing to do with this. Flaky test?

auvipy commented:
seems unit tests are failing

palfrey commented on t/unit/app/test_app.py:
```diff
@@ -591,8 +591,8 @@ def test_worker_main(self, mocked_celery):
-    def test_config_from_envvar(self):
-        os.environ['CELERYTEST_CONFIG_OBJECT'] = 't.unit.app.test_app'
+    def test_config_from_envvar(self, monkeypatch):
+        monkeypatch.setenv("CELERYTEST_CONFIG_OBJECT", 't.unit.app.test_app')
```
This setting was bleeding through into other tests

palfrey commented:
> seems unit tests are failing

Yeah. Test was fine on it's own, but the other tests bled through and messed around with the settings, which made it get stuck. I've done some cleanup to reduce that, and it's certainly able to pass with the other earlier tests locally now.

auvipy commented:
can you recheck why windows builds are failing?

palfrey commented:
> can you recheck why windows builds are failing?

It looked like it was complaining that my test left a thread running afterwards, but that shouldn't be the case. I haven't got a suitable Windows machine to test with so can't even locally figure out why it only happens there.

auvipy commented:
t/unit/bin/test_worker.py::test_cli 
  Error: The action has timed out.

should we try increase the timeout duration?

palfrey commented:
> t/unit/bin/test_worker.py::test_cli Error: The action has timed out.
> 
> should we try increase the timeout duration?

I don't think it's not waiting long enough, I think something is hanging. I've added `--capture=no` to the test args so that we can see the worker config (which might indicate where the problem is) and switched on the solo pool for the test worker to try and remove any dangling thread.

auvipy commented:
good job sir!
